scripts/pylib/pytest-twister-harness/src/twister_harness/device/device_adapter.py

```diff
@@ -69,14 +70,23 @@ def launch(self) -> None:
 
         if self.device_config.type != 'hardware':
             self._flash_and_run()
+            self._device_run.set()
+            self._start_reader_thread()
+            self.connect()
+            return
 
         self._device_run.set()
         self._start_reader_thread()
-        self.connect()
 
-        if self.device_config.type == 'hardware':
+        if self.device_config.flash_before:
+            # For hardware devices with shared USB or software USB, connect after flashing.
+            # Retry for up to 10 seconds for USB-CDC based devices to enumerate.
+            self._flash_and_run()
```

Comment:
flash and run, then open the series, the problem is how do you know there is no log missing?

Reply:
You are correct, but that is the nature of the problem this is solving. When the USB is provided by the app you are testing, the port will not be available until after the firmware is flashed and running.

To mitigate this, I've used a polling loop to establish the connection with a short (100 ms) delay between failed attempts. Also, if the shell is the only logging backend, I don't think any logs will be dropped as they are enqueued until the backend is initialized.